### PR TITLE
Task 70.

### PR DESCRIPTION
**Pull request by AllaKashevarova**
Implemented functionallity to take screenshots on test failure

### CODE REVIEW

File: src/test/java/com/coherent/task60/TestResultLoggerExtension.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class TestResultLoggerExtension implements TestWatcher, AfterAllCallback {
+    private List<TestResultStatus> testResultsStatus = new ArrayList<>();
+    private ScreenshotHelper screenshotHelper = new ScreenshotHelper();
+    private LoginPage loginPage = new LoginPage();
+    private PropertiesHelper propertiesHelper = new PropertiesHelper();
+    private String path = propertiesHelper.propertiesReader("error.screenshots");
```
Comment (KirKarpov) [inline src/test/java/com/coherent/task60/TestResultLoggerExtension.java]: From those variables you need only screenshotHelper

File: src/test/java/com/coherent/task60/TestResultLoggerExtension.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    // @Override
+    public void afterAll(ExtensionContext extensionContext) throws Exception {
+//        Map<TestResultStatus, Long> summary = testResultsStatus.stream()
+//                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+
+    }
+
+    private enum TestResultStatus {
+         FAILED
+    }
```
Comment (KirKarpov) [inline src/test/java/com/coherent/task60/TestResultLoggerExtension.java]: You can remove this enum

File: src/test/java/com/coherent/task60/TestResultLoggerExtension.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    @Override
+    public void testAborted(ExtensionContext context, Throwable cause) {
+        TestWatcher.super.testAborted(context, cause);
+    }
+
+    @Override
+    public void testFailed(ExtensionContext context, Throwable cause) {
+        System.out.println("Test Failed for test {}: ");
+
+        screenshotHelper.takeScreenshot(loginPage.getDriver(), path);
```
Comment (KirKarpov) [inline src/test/java/com/coherent/task60/TestResultLoggerExtension.java]: Better will be to get driver from singleton class

File: src/test/java/com/coherent/task60/LoginPage.java [hunk 4]
```diff
@@ -36,7 +31,12 @@ public By getPasswordField() {
         return passwordField;
     }
 
+    public WebDriver getDriver() {
+        return driver;
+    }
```
Comment (KirKarpov) [inline src/test/java/com/coherent/task60/LoginPage.java]: This method should be removed. If you need to get driver you can call method from singleton class

**Review by KirKarpov — CHANGES_REQUESTED** [review]

File: src/test/java/com/coherent/task60/LoginTest.java [hunk 5]
```diff
@@ -1,30 +1,38 @@
...
 import org.junit.jupiter.api.Test;
-
+import org.junit.jupiter.api.extension.ExtendWith;
 import java.io.IOException;
 
+@ExtendWith(TestResultLoggerExtension.class)
 public class LoginTest {
-    PropertiesHelper propertiesHelper = new PropertiesHelper();
+    private PropertiesHelper propertiesHelper = new PropertiesHelper();
+    private ScreenshotHelper screenshotHelper = new ScreenshotHelper();
```
Comment (KirKarpov) [inline src/test/java/com/coherent/task60/LoginTest.java]: Do you use screenshotHelper and propertiesHelper in this class?

File: src/test/java/com/coherent/task60/LoginTest.java [hunk 6]
```diff
@@ -1,30 +1,38 @@
...
+    private LoginPage loginPage = new LoginPage();
     private String userName;
+    private String path;
 
     {
         userName = propertiesHelper.propertiesReader("user.name");
+        path = propertiesHelper.propertiesReader("error.screenshots");
```
Comment (KirKarpov) [inline src/test/java/com/coherent/task60/LoginTest.java]: I don't see where  path is used

File: src/test/java/com/coherent/task60/LogoutTest.java [hunk 7]
```diff
@@ -1,28 +1,32 @@
...
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
 import org.openqa.selenium.WebDriver;
-
 import java.io.IOException;
 
+@ExtendWith(TestResultLoggerExtension.class)
 public class LogoutTest {
+    private ScreenshotHelper screenshotHelper = new ScreenshotHelper();
+    private PropertiesHelper propertiesHelper = new PropertiesHelper();
```
Comment (KirKarpov) [inline src/test/java/com/coherent/task60/LogoutTest.java]: ScreenshotHelper and PropertiesHelper looks like don't use in the test

**Review by KirKarpov — CHANGES_REQUESTED** [review]

File: src/test/java/com/coherent/task60/TestResultLoggerExtension.java (hunk 1 above)
Comment (AllaKashevarova) [reply src/test/java/com/coherent/task60/TestResultLoggerExtension.java]: Sure. Just forgot to remove them

File: src/test/java/com/coherent/task60/TestResultLoggerExtension.java (hunk 2 above)
Comment (AllaKashevarova) [reply src/test/java/com/coherent/task60/TestResultLoggerExtension.java]: done

File: src/test/java/com/coherent/task60/LoginPage.java (hunk 4 above)
Comment (AllaKashevarova) [reply src/test/java/com/coherent/task60/LoginPage.java]: Got it. Removed

File: src/test/java/com/coherent/task60/LoginTest.java (hunk 5 above)
Comment (AllaKashevarova) [reply src/test/java/com/coherent/task60/LoginTest.java]: screenshotHelper - nope. Removed. But I need propertiesHelper if I want to run the test that doesn't fail

File: src/test/java/com/coherent/task60/LoginTest.java (hunk 6 above)
Comment (AllaKashevarova) [reply src/test/java/com/coherent/task60/LoginTest.java]: removed

File: src/test/java/com/coherent/task60/LogoutTest.java (hunk 7 above)
Comment (AllaKashevarova) [reply src/test/java/com/coherent/task60/LogoutTest.java]: Agree, removed

File: src/test/java/com/coherent/task60/TestResultLoggerExtension.java (hunk 3 above)
Comment (AllaKashevarova) [reply src/test/java/com/coherent/task60/TestResultLoggerExtension.java]: done

**Review by KirKarpov — APPROVED** [review]

**Pull request merged**